### DISCUSSION TITLE
Best practices for uploading texture data: writeTexture vs mappedBuffer -> copyBufferToTexture?

### DISCUSSION THREAD

**Discussion by mmgeorge** [gpuweb/gpuweb#4733] [Q&A] [answered]
Hi all, trying to understand what the best practice for uploading texture data every frame is. In particular, it seems like using `writeTexture` can very negatively impact performance vs using a mapped staging buffer? If I [compare uploading a 4K texture every frame ](https://github.com/mmgeorge/wgpu-texture-upload-comparison/blob/master/index.js#L157-L194) with the two on chrome & a 1080 Ti, I'm getting around 36 FPS with writeTexture vs 144 FPS when using a staging buffer -> `copyBufferToTexture`. 

So if we want to maximize upload throughput, probably we should ignore `writeTexture` all-together & be converting images to byte arrays in a worker, and then sending that over via buffers? Anyone else have any other approaches they've been experimenting with?

**Comment by greggman** [comment, marked as answer]
It depends on what you're trying to do. For video the arguably fastest way is to use `importExternalTexture` [(*)](https://webgpufundamentals.org/webgpu/lessons/webgpu-textures-external-video.html) since it's designed (but not guaranteed) to be zero copy.

Otherwise, I'd expect keeping mapped buffers around and using one to fill with data should fastest.

**Reply by greggman** [reply]
I'm getting ~600fps on a M1 Mac

https://jsgist.org/?src=ae8b00913b4e182a52e4769b733f3695
